frontend/ui/widget/qrwidget.lua

```diff
@@ -0,0 +1,38 @@
+--[[
+QRWidget shows a QR code for a given text.
+]]
+
+local ImageWidget = require("ui/widget/imagewidget")
+local RenderImage = require("ui/renderimage")
+local Widget = require("ui/widget/widget")
+local logger = require("logger")
+local qrencode = require("qrencode")
+
+local QRWidget = Widget:new{
+    scale_factor = 0,
+    text = ""
+    -- see ImageWidget for other options.
+}
+
+function QRWidget:init()
+    local ok, ret = qrencode.qrcode(self.text)
+    if not ok then
+        logger.info("QRWidget: failed to generate QR code.")
+        return
+    else
+        -- We generate pbm data
+        local qr = {"P1\n", #ret[1], " ", #ret, "\n"}
+        for _, col in ipairs(ret) do
+            for _, lgn in ipairs(col) do
+                table.insert(qr, lgn > 0 and "1 " or "0 ")
```

Comment:
You could arguably generate a 4x4 square instead of single pixels to alleviate the "smooth scaling" issue.

Reply:
Just did a 2x2…

Reply:
(Might require some creative intermediary format better suited to that than a Lua table; I'd go with a C char array that can be properly addresses via pointers, because that's what I'm familiar with ^^).

Reply:
I say familiar, because that's roughly what happens here, in FBInk: https://github.com/NiLuJe/FBInk/blob/d8a9e81335fa57eaf31c228456e3b5e4c5f1ffb5/fbink.c#L1433-L1504

Reply:
Or maybe for something that square'ish, you might not need any image: you could just have a widget that itself just paint black rects on its area with something like:
`bb:paintRect(x, y, w, h, Blitbuffer.COLOR_BLACK)`

Reply:
Yep, that's what I was hinting at ;).

Reply:
I'd like to use BlitBuffer methods, but I don't really understand them (my programming skills aren't *that* good…). Moreover, I thought `QRMessage` could evolve towards something that would allow saving the result as a file, in which case the intermediate image representation would be useful.

Reply:
Would mostly be just some Lua fun :) Like:
```lua
function QRWidget:paintTo(bb, x, y)
  -- x/y are the top left position of the top/left of widget in the screen bb
  local grid_size = 64 --dunno what it is really :)
  local sq_size = square_size / grid_size -- square_size is your original "dim"
         -- so best to ensure before that your dim/square_size is a multiple of 64
  for sy=1, #grid_size do
    local sqy = sy*sq_size
    for sx=1, #grid_size do
      if that square is black then
        local sqx = sx*sq_size
        bb:paintRect(x+sqx, y+sqy, sq_size, sq_size, Blitbuffer.COLOR_BLACK)
      end
    end
  end
end
```
(You could then extract that bb from the Screen bb and writePNG() it if you want to save it as a file - or use an intermediate bb here to draw the square to, and at the end write that bb to the screen bb.)

Reply:
You could still feed a BB to RenderImage (I don't recall which method handles that exactly, but this is what the FM does for cover thumbnails, they're stored as raw BBs in the DB).

Basically, start with a w*h*scale BB, fill it white, and then for each input pixel, do a paintRect of scale by scale @ appropriate coordinates. You'll probably need a nested loop, the outer on (input) rows, the inner on (input) columns.
Scaled y == iy*scale (with iy being the input row, starting at row 0)
Scaled x == ix*scale (with ix being the input column, starting at col 0)